pkg/driver/trace_driver.go

```diff
@@ -236,177 +154,117 @@ func (d *Driver) invokeFunction(metadata *InvocationMetadata) {
+			log.Debugf("Test mode invocation fired - ID = %s.\n", invocationID)
+
+			recordOutputChannel <- &mc.ExecutionRecord{
+				ExecutionRecordBase: mc.ExecutionRecordBase{
+					Phase:        int(currentPhase),
+					InvocationID: invocationID,
+					StartTime:    time.Now().UnixNano(),
+				},
+			}
+
+			successfulInvocations++
+		}
+
+		iatIndex++
+
+		d.announceWarmupEnd(startOfExperiment, &currentPhase)
+
+		// counter updates
+		invocationSinceTheBeginningOfMinute++
+		if iatIndex > minuteIndexEnd {
+			interval = minuteIndexSearch.SearchInterval(iatIndex)
+			if interval != nil { // otherwise, the experiment will terminate in the next for loop iteration
+				minuteIndexEnd, minuteIndex, invocationSinceTheBeginningOfMinute = interval.End, interval.Value, 0
 			}
-			numberOfIssuedInvocations++
-			invocationIndex++
 		}
 	}
 
 	waitForInvocations.Wait()
 
 	log.Debugf("All the invocations for function %s have been completed.\n", function.Name)
-	announceFunctionDone.Done()
 
 	atomic.AddInt64(totalSuccessful, successfulInvocations)
 	atomic.AddInt64(totalFailed, failedInvocations)
-	atomic.AddInt64(totalIssued, numberOfIssuedInvocations)
+	atomic.AddInt64(totalIssued, int64(iatIndex))
 }
 
-func (d *Driver) proceedToNextMinute(function *common.Function, minuteIndex *int, invocationIndex *int, startOfMinute *time.Time,
-	skipMinute bool, currentPhase *common.ExperimentPhase, failedInvocationByMinute []int64, previousIATSum *int64) bool {
-
-	if d.Configuration.TraceGranularity == common.MinuteGranularity {
-		if !isRequestTargetAchieved(function.InvocationStats.Invocations[*minuteIndex], *invocationIndex, common.RequestedVsIssued) {
-			// Not fatal because we want to keep the measurements to be written to the output file
-			log.Warnf("Relative difference between requested and issued number of invocations is greater than %.2f%%. Terminating function driver for %s!\n", common.RequestedVsIssuedTerminateThreshold*100, function.Name)
-
-			return true
-		}
-
-		for i := 0; i <= *minuteIndex; i++ {
-			notFailedCount := function.InvocationStats.Invocations[i] - int(atomic.LoadInt64(&failedInvocationByMinute[i]))
-			if !isRequestTargetAchieved(function.InvocationStats.Invocations[i], notFailedCount, common.IssuedVsFailed) {
-				// Not fatal because we want to keep the measurements to be written to the output file
-				log.Warnf("Percentage of failed request is greater than %.2f%%. Terminating function driver for %s!\n", common.FailedTerminateThreshold*100, function.Name)
-
-				return true
-			}
-		}
-	}
-
-	*minuteIndex++
-	*invocationIndex = 0
-	*previousIATSum = 0
-
-	if d.Configuration.WithWarmup() && *minuteIndex == (d.Configuration.LoaderConfiguration.WarmupDuration+1) {
+func (d *Driver) announceWarmupEnd(start time.Time, currentPhase *common.ExperimentPhase) {
+	if *currentPhase == common.WarmupPhase && hasMinuteExpired(start) {
```

Comment:
This makes all warmup 1 minute long. I think we can safely revert back to the check for the current minute instead of real minute expiry (btw, I don't think that would work with second granularity and I don't see any particular use of it now).

Reply:
I am actually not sure about this. What about cases where you have 0 invocations during 5 minutes, and warmup was 1 minute. This function is warmed up at t=1min and scale is 0, but if we change to what you suggest, then warm-up minute will be the first minute in which function is invoked.

I would leave this as is, and extend `hasMinuteExpired` to expiration of `WarmupDuration * time.Second`.

Reply:
You can check it before the invocation happens so we will update the phase to record correctly. It might be written to the log a bit late (e.g., at the 5-minute mark in your example), but still, it will happen before the first invocation in the experiment phase.

Reply:
Fixed.